toponetx/utils/normalization.py

```diff
@@ -0,0 +1,383 @@
+    ----------
+    L : scipy.sparse.csr_matrix
+        The Laplacian matrix.
+
+    Returns
+    -------
+    scipy.sparse.csr_matrix
+        The normalized Laplacian matrix.
+
+    Notes
+    -----
+    This function normalizes the Laplacian matrix by dividing it by the largest eigenvalue.
+    """
+    topeigen_val = spl.eigsh(L.asfptype(), k=1, which="LM", return_eigenvectors=False)[
+        0
+    ]
+    return L * (1.0 / topeigen_val)
+
+
+def compute_x_laplacian_normalized_matrix(L, Lx):
+    """
+    Normalize the up or down Laplacians.
+
+    Parameters
+    ----------
+    L : scipy.sparse.csr_matrix
+        The Laplacian matrix.
+    Lx : scipy.sparse.csr_matrix
+        The up or down Laplacian matrix.
+
+    Returns
+    -------
+    scipy.sparse.csr_matrix
+        The normalized up or down Laplacian matrix.
+
+    Notes
+    -----
+    This function normalizes the up or down Laplacian matrices by dividing them
+    by the largest eigenvalue of the Laplacian matrix.
+    """
+    assert L.shape[0] == L.shape[1]
+    topeig = spl.eigsh(L.asfptype(), k=1, which="LM", return_eigenvectors=False)[0]
+    return Lx * (1.0 / topeig)
+
+
+def compute_kipf_adjacency_normalized_matrix(
+    A_opt, add_identity=False, identity_multiplier=1.0
+):
+    """
+    Normalize the adjacency matrix using Kipf's normalization.
+
+    Typically used to normalize adjacency matrices.
+
+    Parameters
+    ----------
+    A_opt : numpy.ndarray
+        The adjacency matrix.
+    add_identity : bool, optional
+        Determines if the identity matrix is to be added to the adjacency matrix.
+    identity_multiplier : float, optional
+        A multiplier of the identity. This parameter is helpful for higher order
+        (co)adjacency matrices where the neighbor of a cell is obtained from multiple sources.
+
+    Returns
+    -------
+    scipy.sparse.csr_matrix
+        The normalized adjacency matrix.
+
+    Notes
+    -----
+    This normalization is based on Kipf's formulation, which computes the row-sums,
+    constructs a diagonal matrix D^(-0.5), and applies the normalization as D^(-0.5) * A_opt * D^(-0.5).
+    """
+    if add_identity:
+        size = A_opt.shape[0]
+        eye = diags(size * [identity_multiplier])
+        A_opt = np.abs(A_opt) + eye
+    else:
+        A_opt = np.abs(A_opt)
+    rowsum = np.array(A_opt.sum(1))
+    r_inv_sqrt = np.power(rowsum, -0.5).flatten()
+    r_inv_sqrt[np.isinf(r_inv_sqrt)] = 0.0
+    r_mat_inv_sqrt = diags(r_inv_sqrt)
+    A_opt_normalized = A_opt.dot(r_mat_inv_sqrt).transpose().dot(r_mat_inv_sqrt)
+
+    return A_opt_normalized
+
+
+def compute_xu_asymmetric_normalized_matrix(B, is_sparse=True):
+    """
+    Compute Xu's normalized asymmetric matrix.
+
+    Typically used to normalize boundary operators.
+
+    Parameters
+    ----------
+    B : numpy.ndarray or scipy.sparse.csr_matrix
+        The asymmetric matrix.
+    is_sparse : bool, optional
+        If True, treat B as a sparse matrix.
```

Comment:
I wonder whether that parameter is actually useful? Just test whether `B` is sparse ([`scipy.sparse.issparse`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.issparse.html#scipy.sparse.issparse)) should be sufficient, right? Or is there ever a case where `B` is of sparse type but you want to set `is_sparse=False`?